File: llm_arena/workflows/testing.py
"""Automatic test generation, execution and fix loop.

Developer model writes unit + integration tests, reviewer writes edge-case,
error-handling and (where meaningful) performance tests. Failing tests are
sent back to both models for analysis, the developer fixes code (or a wrong
test), tests are re-run – up to ``max_fix_iterations`` times.
"""

from __future__ import annotations

import time
import uuid

from .. import prompts, sandbox
from ..textutil import clip, extract_files, is_test_file, safe_path
from .common import StepFailed, WorkflowContext, other

FAILING = ("failed", "error", "timeout")


class ExecutionDisabled(StepFailed):
    def __init__(self) -> None:
        super().__init__("tests", {"kind": "execution_disabled", "label": "Kódfuttatás tiltva",
                                   "message": "A generált kód futtatása ki van kapcsolva (Beállítások → "
                                              "Kódfuttatás engedélyezése)."})


def roles(ctx: WorkflowContext) -> tuple[str, str]:
    dev = ctx.settings.get("developer", "A")
    return dev, other(dev)


def code_listing(files: dict[str, str], budget: int, *, tests: bool | None = None) -> str:
    """Files as FILE: blocks; ``tests`` None = all, True = only tests, False = only code."""
    names = [n for n in sorted(files) if tests is None or is_test_file(n) == tests]
    if not names:
        return "(nincs fájl)"
    per = max(1500, budget // max(1, len(names)))
    parts = []
    for n in names:
        lang = "python" if n.endswith(".py") else ""
        parts.append(f"FILE: {n}\n```{lang}\n{clip(files[n], per)}\n```")
    return "\n\n".join(parts)


def failures_text(run: dict, limit: int = 12, tb_chars: int = 1500) -> str:
    failing = [t for t in run.get("tests", []) if t["status"] in FAILING]
    parts = []
    for t in failing[:limit]:
        parts.append(f"- {t['id']} [{t['status']}] ({t['file']})\n```\n{clip(t['traceback'] or t['message'], tb_chars)}\n```")
    if len(failing) > limit:
        parts.append(f"... és további {len(failing) - limit} sikertelen teszt")
    if run.get("error"):
        parts.append(f"Futtatási hiba: {run['error']}")
    if run.get("timed_out"):
        parts.append("A tesztfuttatás időtúllépés miatt leállt.")
    if not failing and run.get("stderr"):
        parts.append("stderr:\n```\n" + clip(run["stderr"], 2000) + "\n```")
    return "\n".join(parts) or "(nincs részlet)"


def _state(ctx: WorkflowContext) -> dict:
    return ctx.store.project["testing"]


def _files(ctx: WorkflowContext) -> dict[str, str]:
    with ctx.store.lock:
        return dict(ctx.store.project["code"]["files"])


def _accept_files(files: dict[str, str]) -> dict[str, str]:
    return {p: c for n, c in files.items() if (p := safe_path(n))}


# --------------------------------------------------------------------------- #
def generate_tests(ctx: WorkflowContext, requirements: str) -> dict[str, str]:
    code = {k: v for k, v in _files(ctx).items() if not is_test_file(k)}
    if not code:
        raise StepFailed("tests", {"kind": "no_code", "label": "Nincs kód", "message": "Nincs tesztelhető kód."})
    dev, rev = roles(ctx)
    ctx.progress(0.05, f"Tesztek generálása (LLM {dev}: unit+integrációs, LLM {rev}: edge/hiba/teljesítmény)")

    def ask(slot: str, template: str, title: str, system: str):
        text = ctx.fmt(template, requirements=ctx.clip_for(slot, requirements, 0.2),
                       code=code_listing(code, ctx.budget(slot, 0.6), tests=False))
        return ctx.call(slot, text, role="tester", stage="tests", title=title, system=system)

    results = ctx.parallel({
        dev: lambda: ask(dev, prompts.TESTS_DEV, "Unit + integrációs tesztek", ctx.fmt(prompts.DEV_SYSTEM)),
        rev: lambda: ask(rev, prompts.TESTS_REV, "Edge-case, hibakezelési, teljesítménytesztek",
                         ctx.fmt(prompts.REVIEWER_SYSTEM)),
    })
    generated: dict[str, str] = {}
    by_slot: dict[str, list[str]] = {}
    msg_ids = []
    for slot, (msg, err) in results.items():
        if err is not None:
            ctx.log("warning", f"LLM {slot} tesztgenerálása sikertelen: {err}")
            continue
        msg_ids.append(msg["id"])
        default = "test_unit.py" if slot == dev else "test_edge_cases.py"
        files = {k: v for k, v in _accept_files(extract_files(msg["content"], default_name=default)).items()}
        # Everything a tester returns is a test file; rename accidental non-test names.
        clean = {}
        for name, content in files.items():
            if not is_test_file(name):
                name = f"test_{slot.lower()}_{name.rsplit('/', 1)[-1]}"
            clean[name] = content
        by_slot[slot] = sorted(clean)
        generated.update(clean)
    if not generated:
        raise StepFailed("tests", {"kind": "no_tests", "label": "Nincs teszt",
                                   "message": "Egyik modell sem adott értelmezhető tesztfájlt."})
    ctx.store.set_code(generated, source="tests", note="Generált tesztek: " + ", ".join(sorted(generated)))
    with ctx.store.mutate():
        _state(ctx)["generated"] = {"by_slot": by_slot, "messages": msg_ids, "created": time.time()}
    ctx.state_changed("code")
    ctx.state_changed("testing")
    ctx.log("info", f"{len(generated)} tesztfájl generálva: {', '.join(sorted(generated))}")
    return generated


def run_suite(ctx: WorkflowContext, iteration: int = 0, label: str = "") -> dict:
    s = ctx.settings
    if not s.get("allow_code_execution"):
        raise ExecutionDisabled()
    ctx.check()
    files = _files(ctx)
    ctx.progress(ctx.job.progress, label or f"Tesztek futtatása ({iteration}. iteráció)")
    report = sandbox.run_tests(files, timeout=int(s.get("test_timeout", 60)))
    with ctx.store.lock:
        versions = ctx.store.project["code"]["versions"]
        version = versions[-1]["version"] if versions else 0
    run = {"id": uuid.uuid4().hex[:8], "iteration": iteration, "created": time.time(),
           "code_version": version, **report}
    with ctx.store.mutate():
        _state(ctx)["runs"].append(run)
    sm = run["summary"]
    ctx.log("info" if sm.get("all_passed") else "warning",
            f"Teszteredmény (v{version}): {sm.get('passed', 0)}/{sm.get('total', 0)} sikeres, "
            f"{sm.get('failing', 0)} sikertelen{' – IDŐTÚLLÉPÉS' if run['timed_out'] else ''}")
    ctx.state_changed("testing")
    return run


def _analyse(ctx: WorkflowContext, requirements: str, run: dict, iteration: int) -> tuple[list[dict], list[str]]:
    dev, rev = roles(ctx)
    files = _files(ctx)
    fails = failures_text(run)

    def ask(slot: str):
        text = ctx.fmt(prompts.TEST_ANALYSIS, requirements=ctx.clip_for(slot, requirements, 0.15),
                       code=code_listing(files, ctx.budget(slot, 0.5)), failures=clip(fails, ctx.budget(slot, 0.25)))
        return ctx.call_json(slot, text, role="analyst", stage="test_analysis", round=iteration,
                             title=f"Hibaelemzés – {iteration}. iteráció", temperature=0.2,
                             system=ctx.fmt(prompts.REVIEWER_SYSTEM if slot == rev else prompts.DEV_SYSTEM))

    results = ctx.parallel({rev: lambda: ask(rev), dev: lambda: ask(dev)})
    findings: list[dict] = []
    msg_ids: list[str] = []
    for slot in (rev, dev):
        res, err = results.get(slot, (None, None))
        if err is not None or not res:
            continue
        msg, data = res
        msg_ids.append(msg["id"])
        for f in (data or {}).get("failures") or []:
            if isinstance(f, dict):
                findings.append({"by": slot, "test": str(f.get("test") or ""), "root_cause": str(f.get("root_cause") or ""),
                                 "fix_target": str(f.get("fix_target") or "code"),
                                 "location": str(f.get("location") or ""), "suggestion": str(f.get("suggestion") or "")})
    if not msg_ids:
        raise StepFailed("test_analysis", {"kind": "analysis_failed", "label": "Elemzés sikertelen",
                                           "message": "Egyik modell sem tudta elemezni a hibákat."})
    return findings, msg_ids


def _analysis_text(findings: list[dict]) -> str:
    if not findings:
        return "(a modellek nem adtak strukturált elemzést – a hibaüzenetek alapján javíts)"
    return "\n".join(f"- [{f['by']}] {f['test']}: {f['root_cause']} → javítandó: {f['fix_target']} "
                     f"({f['location']}); javaslat: {f['suggestion']}" for f in findings)


def fix_loop(ctx: WorkflowContext, requirements: str, max_iterations: int | None = None) -> dict:
    s = ctx.settings
    max_iterations = int(max_iterations if max_iterations is not None else s.get("max_fix_iterations", 3))
    dev, _ = roles(ctx)
    loop_id = uuid.uuid4().hex[:8]
    run = run_suite(ctx, 0, "Tesztek futtatása (kiinduló állapot)")
    iteration = 0
    while not run["summary"].get("all_passed") and iteration < max_iterations:
        iteration += 1
        ctx.check()
        base = 0.2 + 0.7 * (iteration - 1) / max(1, max_iterations)
        failing_ids = [t["id"] for t in run["tests"] if t["status"] in FAILING]
        rec = {"loop": loop_id, "iteration": iteration, "created": time.time(), "failing_before": failing_ids,
               "findings": [], "messages": [], "changed_files": [], "fixed": [], "failing_after": [],
               "status": "running"}
        with ctx.store.mutate():
            _state(ctx)["iterations"].append(rec)
        ctx.state_changed("testing")

        ctx.progress(base, f"{iteration}. javító iteráció – hibák elemzése")
        findings, msg_ids = _analyse(ctx, requirements, run, iteration)
        with ctx.store.mutate():
            rec["findings"] = findings
            rec["messages"] += msg_ids
        ctx.state_changed("testing")

        ctx.progress(base + 0.1, f"{iteration}. javító iteráció – javítás (LLM {dev})")
        files = _files(ctx)
        text = ctx.fmt(prompts.TEST_FIX, requirements=ctx.clip_for(dev, requirements, 0.12),
                       code=code_listing(files, ctx.budget(dev, 0.5)),
                       failures=clip(failures_text(run), ctx.budget(dev, 0.18)),
                       analysis=clip(_analysis_text(findings), ctx.budget(dev, 0.15)),
                       code_rules=ctx.fmt(prompts.CODE_RULES))
        msg = ctx.call(dev, text, role="developer", stage="test_fix", round=iteration,
                       title=f"Javítás – {iteration}. iteráció", system=ctx.fmt(prompts.DEV_SYSTEM))
        with ctx.store.mutate():
            rec["messages"].append(msg["id"])
        changed = {n: c for n, c in _accept_files(extract_files(msg["content"], default_name="")).items()
                   if n and files.get(n) != c}
        changed.pop("", None)
        if not changed:
            with ctx.store.mutate():
                rec["status"] = "no_change"
            ctx.log("warning", f"{iteration}. iteráció: a javítás nem módosított egyetlen fájlt sem – a ciklus leáll.")
            ctx.state_changed("testing")
            break
        version = ctx.store.set_code(changed, source=f"test_fix_{iteration}",
                                     note=f"Tesztjavítás {iteration}. iteráció")
        with ctx.store.mutate():
            rec["changed_files"] = sorted(changed)
            rec["code_version"] = version["version"]
        ctx.state_changed("code")

        new_run = run_suite(ctx, iteration, f"{iteration}. iteráció – tesztek újrafuttatása")
        status = {t["id"]: t["status"] for t in new_run["tests"]}
        fixed = [tid for tid in failing_ids if status.get(tid) == "passed"]
        with ctx.store.mutate():
            rec["fixed"] = fixed
            rec["failing_after"] = [t["id"] for t in new_run["tests"] if t["status"] in FAILING]
            rec["status"] = "done"
            for tid in fixed:
                f = next((f for f in findings if f["test"] and (f["test"] in tid or tid.endswith(f["test"]))), None)
                _state(ctx)["fixed_bugs"].append({
                    "test": tid, "iteration": iteration, "code_version": version["version"],
                    "root_cause": (f or {}).get("root_cause", ""), "fix_target": (f or {}).get("fix_target", ""),
                    "by": dev, "loop": loop_id,
                })
        ctx.state_changed("testing")
        run = new_run

    report = build_report(ctx, run, iteration, max_iterations, loop_id)
    ctx.progress(1.0, "Tesztelés kész" + (" – minden teszt sikeres" if report["all_passed"] else ""))
    return report


def build_report(ctx: WorkflowContext, run: dict, iterations: int, max_iterations: int,
                 loop_id: str | None = None) -> dict:
    sm = run["summary"]
    with ctx.store.lock:
        t = _state(ctx)
        fixed = [b for b in t["fixed_bugs"] if loop_id is None or b.get("loop") == loop_id]
        code_files = dict(ctx.store.project["code"]["files"])
        versions = ctx.store.project["code"]["versions"]
    remaining = [f"{x['id']}: {x['message'] or x['status']}" for x in run["tests"] if x["status"] in FAILING]
    if run.get("error"):
        remaining.append(run["error"])
    report = {
        "created": time.time(),
        "passed": sm.get("passed", 0),
        "failed": sm.get("failing", 0),
        "total": sm.get("total", 0),
        "skipped": sm.get("skipped", 0),
        "all_passed": bool(sm.get("all_passed")),
        "by_category": sm.get("by_category", {}),
        "iterations": iterations,
        "max_iterations": max_iterations,
        "fixed_bugs": fixed,
        "remaining_issues": remaining,
        "final_version": versions[-1]["version"] if versions else 0,
        "final_files": sorted(code_files),
        "last_run": run["id"],
    }
    with ctx.store.mutate():
        _state(ctx)["report"] = report
    ctx.state_changed("testing")
    return report


def summary_text(report: dict | None) -> str:
    if not report:
        return "(a tesztek nem futottak)"
    lines = [f"Sikeres: {report['passed']} / {report['total']}, sikertelen: {report['failed']}, "
             f"javító iterációk: {report['iterations']}, javított hibák: {len(report['fixed_bugs'])}"]
    if report["remaining_issues"]:
        lines.append("Fennmaradó problémák:\n" + "\n".join(f"- {r}" for r in report["remaining_issues"][:10]))
    return "\n".join(lines)

File: llm_arena/static/js/views/testing.js
// Automatic testing: generate → run → analyse → fix loop, with results.
import { api } from "../api.js";
import { state, trackJob, runningJobs, refresh } from "../state.js";
import { esc, toast, runningBanner, statusBadge, fmtDate, messageCard, lastFinishedError, slotBadge } from "../ui.js";

let root;
let showTb = new Set();
const CAT = { unit: "Unit", integration: "Integrációs", edge_case: "Edge-case", error_handling: "Hibakezelés", performance: "Teljesítmény" };

function testsTable(run) {
  if (!run) return '<div class="empty">Még nem futottak tesztek.</div>';
  if (!run.tests.length) return `<div class="errbox">${esc(run.error || "Nincs teszteredmény.")}</div>${run.stderr ? `<pre class="tb">${esc(run.stderr)}</pre>` : ""}`;
  return `<table class="tbl"><thead><tr><th>Állapot</th><th>Kategória</th><th>Teszt</th><th>Fájl</th><th>Idő</th><th>Üzenet</th></tr></thead><tbody>
    ${run.tests.map((t) => `<tr class="clickable" data-tb="${esc(t.id)}"><td>${statusBadge(t.status)}</td><td>${esc(CAT[t.category] || t.category)}</td>
      <td class="mono">${esc(t.id)}</td><td class="mono">${esc(t.file)}</td><td>${t.duration != null ? (t.duration * 1000).toFixed(0) + " ms" : "–"}</td>
      <td>${esc(t.message)}</td></tr>
      ${showTb.has(t.id) && t.traceback ? `<tr><td colspan="6"><pre class="tb">${esc(t.traceback)}</pre></td></tr>` : ""}`).join("")}
  </tbody></table>`;
}

function iterationsHtml(its) {
  if (!its.length) return '<div class="hint">Még nem volt javító iteráció.</div>';
  return [...its].reverse().slice(0, 20).map((it) => `<div class="card">
    <div class="card-head"><h3>${it.iteration}. iteráció</h3>${statusBadge(it.status)}
      <span class="hint">${fmtDate(it.created)} · előtte ${it.failing_before.length} sikertelen → utána ${it.failing_after.length}</span>
      ${it.code_version ? `<span class="badge">kód v${it.code_version}</span>` : ""}</div>
    ${it.findings.length ? `<table class="tbl"><thead><tr><th>Elemző</th><th>Teszt</th><th>Gyökérok</th><th>Javítandó</th><th>Javaslat</th></tr></thead><tbody>
      ${it.findings.map((f) => `<tr><td>${slotBadge(f.by)}</td><td class="mono">${esc(f.test)}</td><td>${esc(f.root_cause)}</td><td>${esc(f.fix_target)}</td><td>${esc(f.suggestion)}</td></tr>`).join("")}</tbody></table>` : ""}
    ${it.changed_files.length ? `<div class="hint mt">Módosított fájlok: <span class="mono">${esc(it.changed_files.join(", "))}</span></div>` : ""}
    ${it.fixed.length ? `<div class="mt">✅ Javítva: <span class="mono">${esc(it.fixed.join(", "))}</span></div>` : ""}
    <details class="plain mt"><summary>Modellválaszok (${it.messages.length})</summary><div class="stepper mt">${it.messages.map((id) => messageCard(id)).join("")}</div></details>
  </div>`).join("");
}

async function start(action) {
  try {
    const body = action === "loop" ? { max_iterations: +root.querySelector("#t-iter").value } : {};
    trackJob((await api.testing(action, body)).job);
  } catch (e) { toast(e.message, "error"); }
}

export default {
  mount(el) {
    root = el;
    root.innerHTML = `<h1>Automatikus tesztelés</h1>
      <p class="subtitle">A fejlesztő modell unit + integrációs, a reviewer edge-case, hibakezelési és teljesítménytesztet ír. Bukás esetén mindkét modell elemzi a hibát, a fejlesztő javít, a tesztek újrafutnak – a megadott maximális iterációig.</p>
      <div id="t-exec"></div>
      <div class="card"><div class="row">
        <label class="field" style="max-width:180px"><span>Max. javító iteráció</span><input type="number" id="t-iter" min="0" max="10" value="${state.project.settings.max_fix_iterations}"></label>
        <span class="spacer"></span>
        <button class="btn" data-t="generate">🧪 Tesztek generálása</button>
        <button class="btn" data-t="run">▶ Tesztek futtatása</button>
        <button class="btn primary" data-t="loop">⟳ Teszt → javítás ciklus</button>
      </div></div>
      <div id="t-banner"></div><div id="t-body"></div>`;
    root.addEventListener("click", async (e) => {
      const b = e.target.closest("[data-t]");
      if (b) { start(b.dataset.t); return; }
      if (e.target.id === "t-enable") {
        if (!confirm("Engedélyezed a generált kód futtatását ezen a gépen?")) return;
        await api.saveConfig({ settings: { allow_code_execution: true } });
        state.project.settings.allow_code_execution = true;
        refresh(10);
        this.update("project");
        return;
      }
      const tr = e.target.closest("[data-tb]");
      if (tr) { const id = tr.dataset.tb; if (showTb.has(id)) showTb.delete(id); else showTb.add(id); this.update("project"); }
    });
  },
  update(reason) {
    root.querySelector("#t-banner").innerHTML = runningBanner(["testing", "design", "pipeline"]);
    const busy = runningJobs("testing").length || runningJobs("design").length || runningJobs("pipeline").length;
    const t = state.project.testing;
    const hasCode = Object.keys(state.project.code.files).length > 0;
    const allow = state.project.settings.allow_code_execution;
    root.querySelectorAll("[data-t]").forEach((b) => { b.disabled = !!busy || !hasCode || (b.dataset.t !== "generate" && !allow); });
    root.querySelector("#t-exec").innerHTML = allow ? "" : `<div class="warnbox">A generált kód futtatása ki van kapcsolva. A tesztek generálhatók, de nem futtathatók.
      <button class="btn small" id="t-enable">Engedélyezés</button></div>`;
    if (reason === "jobs") return;
    const rep = t.report;
    const last = t.runs[t.runs.length - 1];
    const cats = rep?.by_category || last?.summary?.by_category || {};
    root.querySelector("#t-body").innerHTML = `
      ${lastFinishedError("testing")}
      ${rep ? `<div class="stats">
        <div class="stat ok"><div class="v">${rep.passed}</div><div class="k">Sikeres teszt</div></div>
        <div class="stat ${rep.failed ? "err" : ""}"><div class="v">${rep.failed}</div><div class="k">Sikertelen teszt</div></div>
        <div class="stat"><div class="v">${rep.fixed_bugs.length}</div><div class="k">Javított hiba</div></div>
        <div class="stat ${rep.remaining_issues.length ? "warn" : ""}"><div class="v">${rep.remaining_issues.length}</div><div class="k">Fennmaradó probléma</div></div>
        <div class="stat"><div class="v">${rep.iterations}/${rep.max_iterations}</div><div class="k">Javító iteráció</div></div>
        <div class="stat"><div class="v">v${rep.final_version}</div><div class="k">Végleges kód <a href="#code" data-tab="code" data-goto>→</a></div></div>
      </div>` : hasCode ? "" : '<div class="empty">Még nincs kód. Előbb futtass programtervezést (vagy hozz létre fájlt a Kódnézetben).</div>'}
      ${Object.keys(cats).length ? `<div class="row" style="margin-bottom:12px">${Object.entries(cats).map(([k, v]) => `<span class="badge ${v.passed === v.total ? "ok" : "err"}">${esc(CAT[k] || k)}: ${v.passed}/${v.total}</span>`).join("")}</div>` : ""}
      ${rep && rep.remaining_issues.length ? `<div class="card"><h3>Fennmaradó problémák</h3><ul class="clean">${rep.remaining_issues.map((r) => `<li class="mono">${esc(r)}</li>`).join("")}</ul></div>` : ""}
      ${rep && rep.fixed_bugs.length ? `<div class="card"><h3>Javított hibák</h3><table class="tbl"><thead><tr><th>Teszt</th><th>Iteráció</th><th>Gyökérok</th><th>Javítás helye</th><th>Kódverzió</th></tr></thead><tbody>
        ${rep.fixed_bugs.map((b) => `<tr><td class="mono">${esc(b.test)}</td><td>${b.iteration}</td><td>${esc(b.root_cause)}</td><td>${esc(b.fix_target)}</td><td>v${b.code_version}</td></tr>`).join("")}</tbody></table></div>` : ""}
      <h2>Legutóbbi futtatás ${last ? `<span class="hint">· ${fmtDate(last.created)} · kód v${last.code_version} · ${last.duration}s${last.timed_out ? " · IDŐTÚLLÉPÉS" : ""}</span>` : ""}</h2>
      ${testsTable(last)}
      ${last && (last.stdout || last.stderr) ? `<details class="plain mt"><summary>Kimenet (stdout/stderr)</summary><pre class="tb">${esc(last.stdout)}\n${esc(last.stderr)}</pre></details>` : ""}
      <h2>Javító iterációk</h2>${iterationsHtml(t.iterations)}
      <h2>Futtatási előzmények</h2>
      <table class="tbl"><thead><tr><th>Idő</th><th>Iteráció</th><th>Kód</th><th>Sikeres</th><th>Sikertelen</th><th>Időtartam</th></tr></thead><tbody>
        ${[...t.runs].reverse().slice(0, 30).map((r) => `<tr><td>${fmtDate(r.created)}</td><td>${r.iteration}</td><td>v${r.code_version}</td><td>${r.summary.passed ?? 0}/${r.summary.total ?? 0}</td><td>${r.summary.failing ?? 0}</td><td>${r.duration}s</td></tr>`).join("") || '<tr><td colspan="6" class="hint">nincs</td></tr>'}
      </tbody></table>`;
  },
};
